### PR TITLE
feature/1339_kpis_api_common_ops_outgoing

### PR DESCRIPTION
**Pull request by frbattid**
* Partially implements issue #1339 (outgoing metrics)
* (Unofficial) e2e tests passed:

HDFS sink:

```
{
    "services": {
        "sc_valencia": {
            "subservs": {
                "/gardens": {
                    "incomingTransactionErrors": 0,
                    "incomingTransactionRequestSize": 460,
                    "incomingTransactionResponseSize": 0,
                    "incomingTransactions": 1,
                    "outgoingTransactionErrors": 0,
                    "outgoingTransactionRequestSize": 216,
                    "outgoingTransactionResponseSize": 205,
                    "outgoingTransactions": 1,
                    "serviceTime": 1213.0
                }
            },
            "sum": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 460,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 1,
                "outgoingTransactionErrors": 0,
                "outgoingTransactionRequestSize": 216,
                "outgoingTransactionResponseSize": 205,
                "outgoingTransactions": 1,
                "serviceTime": 1213.0
            }
        }
    },
    "sum": {
        "subservs": {
            "/gardens": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 460,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 1,
                "outgoingTransactionErrors": 0,
                "outgoingTransactionRequestSize": 216,
                "outgoingTransactionResponseSize": 205,
                "outgoingTransactions": 1,
                "serviceTime": 1213.0
            }
        },
        "sum": {
            "incomingTransactionErrors": 0,
            "incomingTransactionRequestSize": 460,
            "incomingTransactionResponseSize": 0,
            "incomingTransactions": 1,
            "outgoingTransactionErrors": 0,
            "outgoingTransactionRequestSize": 216,
            "outgoingTransactionResponseSize": 205,
            "outgoingTransactions": 1,
            "serviceTime": 1213.0
        }
    }
}
```

CKAN sink:

```
{
    "services": {
        "sc_valencia": {
            "subservs": {
                "/gardens": {
                    "incomingTransactionErrors": 0,
                    "incomingTransactionRequestSize": 460,
                    "incomingTransactionResponseSize": 0,
                    "incomingTransactions": 1,
                    "outgoingTransactionErrors": 0,
                    "outgoingTransactionRequestSize": 322,
                    "outgoingTransactionResponseSize": 414,
                    "outgoingTransactions": 1,
                    "serviceTime": 12828.0
                }
            },
            "sum": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 460,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 1,
                "outgoingTransactionErrors": 0,
                "outgoingTransactionRequestSize": 322,
                "outgoingTransactionResponseSize": 414,
                "outgoingTransactions": 1,
                "serviceTime": 12828.0
            }
        }
    },
    "sum": {
        "subservs": {
            "/gardens": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 460,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 1,
                "outgoingTransactionErrors": 0,
                "outgoingTransactionRequestSize": 322,
                "outgoingTransactionResponseSize": 414,
                "outgoingTransactions": 1,
                "serviceTime": 12828.0
            }
        },
        "sum": {
            "incomingTransactionErrors": 0,
            "incomingTransactionRequestSize": 460,
            "incomingTransactionResponseSize": 0,
            "incomingTransactions": 1,
            "outgoingTransactionErrors": 0,
            "outgoingTransactionRequestSize": 322,
            "outgoingTransactionResponseSize": 414,
            "outgoingTransactions": 1,
            "serviceTime": 12828.0
        }
    }
}
```

Carto sink:

```
{
    "services": {
        "myservice": {
            "subservs": {
                "/gardens": {
                    "incomingTransactionErrors": 0,
                    "incomingTransactionRequestSize": 1498,
                    "incomingTransactionResponseSize": 0,
                    "incomingTransactions": 2,
                    "outgoingTransactionErrors": 0,
                    "outgoingTransactionRequestSize": 0,
                    "outgoingTransactionResponseSize": 931,
                    "outgoingTransactions": 2,
                    "serviceTime": 13417.0
                }
            },
            "sum": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 1498,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 2,
                "outgoingTransactionErrors": 0,
                "outgoingTransactionRequestSize": 0,
                "outgoingTransactionResponseSize": 931,
                "outgoingTransactions": 2,
                "serviceTime": 13417.0
            }
        }
    },
    "sum": {
        "subservs": {
            "/gardens": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 1498,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 2,
                "outgoingTransactionErrors": 0,
                "outgoingTransactionRequestSize": 0,
                "outgoingTransactionResponseSize": 931,
                "outgoingTransactions": 2,
                "serviceTime": 13417.0
            }
        },
        "sum": {
            "incomingTransactionErrors": 0,
            "incomingTransactionRequestSize": 1498,
            "incomingTransactionResponseSize": 0,
            "incomingTransactions": 2,
            "outgoingTransactionErrors": 0,
            "outgoingTransactionRequestSize": 0,
            "outgoingTransactionResponseSize": 931,
            "outgoingTransactions": 2,
            "serviceTime": 13417.0
        }
    }
}
```

HDFS persistence error, 3 retries:

```
{
    "services": {
        "sc_valencia": {
            "subservs": {
                "/gardens": {
                    "incomingTransactionErrors": 0,
                    "incomingTransactionRequestSize": 460,
                    "incomingTransactionResponseSize": 0,
                    "incomingTransactions": 1,
                    "outgoingTransactionErrors": 3,
                    "outgoingTransactionRequestSize": 0,
                    "outgoingTransactionResponseSize": 0,
                    "outgoingTransactions": 3,
                    "serviceTime": 10052.666666666666
                }
            },
            "sum": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 460,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 1,
                "outgoingTransactionErrors": 3,
                "outgoingTransactionRequestSize": 0,
                "outgoingTransactionResponseSize": 0,
                "outgoingTransactions": 3,
                "serviceTime": 10052.666666666666
            }
        }
    },
    "sum": {
        "subservs": {
            "/gardens": {
                "incomingTransactionErrors": 0,
                "incomingTransactionRequestSize": 460,
                "incomingTransactionResponseSize": 0,
                "incomingTransactions": 1,
                "outgoingTransactionErrors": 3,
                "outgoingTransactionRequestSize": 0,
                "outgoingTransactionResponseSize": 0,
                "outgoingTransactions": 3,
                "serviceTime": 10052.666666666666
            }
        },
        "sum": {
            "incomingTransactionErrors": 0,
            "incomingTransactionRequestSize": 460,
            "incomingTransactionResponseSize": 0,
            "incomingTransactions": 1,
            "outgoingTransactionErrors": 3,
            "outgoingTransactionRequestSize": 0,
            "outgoingTransactionResponseSize": 0,
            "outgoingTransactions": 3,
            "serviceTime": 10052.666666666666
        }
    }
}
```

### CODE REVIEW

File: cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/http/HttpBackendTest.java [hunk 1]
```diff
@@ -187,7 +188,7 @@ public void testCreateJsonResponseNoLocationHeader() {
             throw new AssertionError(e.getMessage());
         } // try catch
         
-        response.addHeader("Content-Type", "text/html");
+        response.addHeader("Content-Type", "application/json");
```
Comment (fgalan) [inline cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/http/HttpBackendTest.java]: This change is related with the metrics functionality addressed by this PR?

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISink.java [hunk 2]
```diff
@@ -443,7 +447,7 @@ protected Accumulator getRollbackedAccumulationForRetry() {
      * Rollbacks the accumulation once more.
      * @param rollbackedAccumulation
      */
-    protected void doRollbackAgain(Accumulator rollbackedAccumulation) {
+    protected void doRollbackAgain(Accumulator rollbackedAccumulation) {
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISink.java]: Extra whitespace to be removed?

File: doc/cygnus-common/installation_and_administration_guide/management_interface_v1.md [hunk 3]
```diff
@@ -1011,6 +1015,8 @@ Additionally, because Cygnus distributes event processing among sources (respons
 * The service time is computed as the total sum of processing times (understanding processing time in this context as the time between even reception in the source and processing in one of the invidual sinks) divided by the number of times the same incoming event is persisted (in other words, it is not divided by the number of incoming transactions). Thus, it could occur N events are persisted M times (M > N), and that impacts in the average service time.
 * The same occurs with the number of transaction errors. Thus, it could be this number is greater than the number of incoming transactions because the events are being processed by 2 or more sinks, and all of them are failing.
 
+Finally, because Cygnus implements a retry mechanism for those persistence operations that failed, it must be noticed the number of outgoing transactions (and its related metrics) may be much higher than the number of icoming transactions (even when a single a sink is used).
```
Comment (fgalan) [inline doc/cygnus-common/installation_and_administration_guide/management_interface_v1.md]: icoming -> incoming

I will add this sentence to illustrate how it works (if it is correct, of course): "Each new retry counts as a new outgoing transaction, e.g. if a persistence operations takes 4 retries (the last being the one that finally have success) this means 4 `outgoingTransactions`, 3 of them `outgoingTransactionError`.

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/http/HttpBackend.java [hunk 4]
```diff
@@ -31,6 +31,7 @@
 import javax.security.auth.Subject;
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/http/HttpBackend.java]: CNR entry about metrics already added, isn't it?

Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/http/HttpBackend.java]: Yes.

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISink.java (hunk 2 above)
Comment (frbattid) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISink.java]: Yes, Telefonica's checkstyle warned me about this.

File: cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/http/HttpBackendTest.java (hunk 1 above)
Comment (frbattid) [inline cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/http/HttpBackendTest.java]: No, it is not. I discovered the test was wrong and fixed it... :)

File: doc/cygnus-common/installation_and_administration_guide/management_interface_v1.md (hunk 3 above)
Comment (frbattid) [inline doc/cygnus-common/installation_and_administration_guide/management_interface_v1.md]: Added in https://github.com/telefonicaid/fiware-cygnus/pull/1385/commits/3ca0165de8d17f323800172de530063b71bc8748

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/http/HttpBackend.java (hunk 4 above)
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/http/HttpBackend.java]: NTC

File: cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/http/HttpBackendTest.java (hunk 1 above)
Comment (fgalan) [inline cygnus-common/src/test/java/com/telefonica/iot/cygnus/backends/http/HttpBackendTest.java]: NTC

File: doc/cygnus-common/installation_and_administration_guide/management_interface_v1.md (hunk 3 above)
Comment (frbattid) [inline doc/cygnus-common/installation_and_administration_guide/management_interface_v1.md]: "icoming" issue fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1385/commits/07a7b2349d27c6801078b2717cdd83c0a9bac56e

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISink.java (hunk 2 above)
Comment (frbattid) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSISink.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1385/commits/07a7b2349d27c6801078b2717cdd83c0a9bac56e

**Comment by fgalan** [comment]
LGTM

**Pull request merged**